File: dart/lib/bank.dart
import 'money.dart';

class Bank {
  Map<String, double> exchangeRates = {};

  Bank();

  addExchangeRate(
    String currencyFrom,
    String currencyTo,
    double rate,
  ) {
    final key = '$currencyFrom->$currencyTo';
    exchangeRates[key] = rate;
  }

  Money? convert(Money money, String currency) {
    if (money.currency == currency) {
      return Money(money.amount, money.currency);
    }

    final key = '${money.currency}->$currency';
    final rate = exchangeRates[key];
    if (rate == null) {
      throw UnsupportedError(key);
    }

    return Money(money.amount * rate, currency);
  }
}

File: typescript/src/bank.ts
import { Money } from './money';
export class Bank {
    exchangeRates: Map<string, number>;

    constructor() {
        this.exchangeRates = new Map();
    }

    addExchangeRate(currencyFrom: string, currencyTo: string, rate: number) {
        const key = `${currencyFrom}->${currencyTo}`;
        this.exchangeRates.set(key, rate);
    }

    convert(money: Money, currency: string): Money {
        if (money.currency === currency) {
            return new Money(money.amount, money.currency);
        }

        const key = `${money.currency}->${currency}`;
        const rate = this.exchangeRates.get(key);

        if (rate === undefined) {
            throw new Error(key);
        }

        return new Money(money.amount * rate, currency);
    }
}